File: libs/power-log-parser/src/lib/parsers/battlegrounds-hero-killed-parser.ts
import { BlockType, CardType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsHeroKilledParser implements ActionParser {
	readonly ParserName = 'BattlegroundsHeroKilledParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, helper: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = helper;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === TagChange &&
			(node.Object as TagChange).Name === (GameTag.ZONE as number) &&
			(node.Object as TagChange).Value === (Zone.GRAVEYARD as number) &&
			this.GameState.CurrentEntities.get((node.Object as TagChange).Entity)?.GetTag(GameTag.CARDTYPE) ===
				(CardType.HERO as number)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const parent = node.Parent;
		if (
			parent == null ||
			!(parent.Object instanceof Action) ||
			(parent.Object as Action).Type !== (BlockType.DEATHS as number)
		) {
			return null;
		}

		if (
			this.GameState.CurrentEntities.get(tagChange.Entity)!.GetController() ===
			this.StateFacade.LocalPlayer!.Id
		) {
			return null;
		}

		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'BATTLEGROUNDS_ENEMY_HERO_KILLED',
				GameEventHelper.CreateProvider(
					'BATTLEGROUNDS_ENEMY_HERO_KILLED',
					this.GameState.CurrentEntities.get(tagChange.Entity)!.CardId,
					this.GameState.CurrentEntities.get(tagChange.Entity)!.GetController(),
					tagChange.Entity,
					this.StateFacade,
					null,
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsHeroKilledParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsHeroKilledParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsHeroKilledParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.ZONE
                && (node.Object as TagChange).Value == (int)Zone.GRAVEYARD
                && GameState.CurrentEntities[(node.Object as TagChange).Entity].GetTag(GameTag.CARDTYPE) == (int)CardType.HERO;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var parent = node.Parent;
            // If it's not in a DEATHS block, it could simply be that the hero has been swapped out
            if (parent == null || !(parent.Object is Action) || (parent.Object as Action).Type != (int)BlockType.DEATHS)
            {
                return null;
            }

            if (GameState.CurrentEntities[tagChange.Entity].GetController() == StateFacade.LocalPlayer.Id)
            {
                return null;
            }

            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                "BATTLEGROUNDS_ENEMY_HERO_KILLED",
                GameEvent.CreateProvider(
                    "BATTLEGROUNDS_ENEMY_HERO_KILLED",
                    GameState.CurrentEntities[tagChange.Entity].CardId,
                    GameState.CurrentEntities[tagChange.Entity].GetController(),
                    tagChange.Entity,
                    StateFacade,
                    null),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
